### PR TITLE
feat(geocoder): implement parallel processing to leverage full QPS

### PR DESCRIPTION
**Pull request by Caerostris**
This PR implements parallel processing for the batch geocoder, and thereby increases the effective QPS significantly.
The implementation of the parallel transform stream has been adapted from [parallel-transform](https://www.npmjs.com/package/parallel-transform) (MIT license).
The parallel transform code has been re-written in ES6, and was changed to provide an inheritance-based API.
It might be better to move `parallelTransform.js` into a separate npm package.
I will add test cases & documentation for `parallelTransform.js` if it withstands an initial review.

### CODE REVIEW

File: src/geocoder.js [hunk 1]
```diff
@@ -68,44 +66,54 @@ export default class Geocoder {
...
+      }, 100);
     }
 
-    this.currentRequests++;
-    this.lastGeocode = now;
+    this.queries++;
+    this.startGeocode(address, resolve, reject);
+  }
 
+  startBucket() {
```
Comment (aemkei) [inline src/geocoder.js:111]: Please add some description + explanations here.

File: src/lib/parallel-transform.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import stream from 'stream';
```
Comment (aemkei) [inline src/lib/parallel-transform.js:1]: This needs some inline docs.

**Comment by Caerostris** [comment]
dcf3009 changed an entry in `.eslintrc`. Travis has decided to ignore these changes for some reason, and eslint is therefore now failing on Travis

**Comment by ro-ka** [comment]
Hmm, wanted to clear the cache but I can’t. Rerunning either way.

There are conflicts with your other branch now, resolve these.

File: src/geocode-stream.js [hunk 3]
```diff
@@ -14,35 +14,32 @@ export default class GeocodeStream extends stream.Transform {
    *                             data item directly.
    */
   constructor(geocoder, stats, accessor = address => address) {
-    super({objectMode: true});
+    super(50, {objectMode: true});
```
Comment (ro-ka) [inline src/geocode-stream.js]: Using a const makes it easier to get what the 50 is about.

File: src/geocoder.js [hunk 4]
```diff
@@ -48,10 +48,8 @@ export default class Geocoder {
     options = Object.assign({}, defaults, options);
     validateOptions(options);
 
-    this.timeBetweenRequests = 20;
-    this.maxRequests = 20;
-    this.lastGeocode = new Date();
-    this.currentRequests = 0;
+    this.qps = 50;
```
Comment (ro-ka) [inline src/geocoder.js]: Use the config introduced in the other PR.

File: src/geocoder.js [hunk 5]
```diff
@@ -68,44 +66,58 @@ export default class Geocoder {
...
-    ) {
+    if (this.queries === -1) {
+      this.startBucket();
+    } else if (this.queries >= this.qps) {
+      // maximum number of queries for this bucket exceeded
       return setTimeout(() => {
-        this.startGeocode(address, resolve, reject);
-      }, this.timeBetweenRequests);
+        this.queueGeocode(address, resolve, reject);
+      }, 100);
```
Comment (ro-ka) [inline src/geocoder.js]: Why 100?

**Comment by ro-ka** [comment]
What speaks against using the existing parallel-transform package?

**Comment by Caerostris** [comment]
The documented API of the existing parallel-transform package only allows the user to pass a callback function. It offers no way to parallelise an instance of `stream.Transform`.
Thus, using the API of the existing parallel-transform package involves removing the inheritance of `stream.Transform` from the `GeocodeStream` class. Doing so breaks the majority of all test cases, and would require a re-write of half the testing framework. It would also introduce a breaking change to the API of Geobatch, as `geoBatch.geocode()` would now return an instance of `ParallelTransform` instead of  `GeocodeStream`.
The changes required to change the API of the parallel-transform package to be based on inheritance instead of callbacks are minimal (~3 lines).

File: src/geocoder.js (hunk 5 above)
Comment (Caerostris) [inline src/geocoder.js]: You are probably right. Changing it to the duration of a bucket should also work and would probably make more sense.
The impact this timeout is minimal in any case. I'll experiment with it 👍

**Comment by ro-ka** [comment]
Okay, got it. Better not break the API.

**Comment by Caerostris** [comment]
In any case, the QPS limitation code has not yet been tested under high load.
This PR should not be merged until it had a change to prove itself with a few thousand requests.
I still wanted to get a PR up to get some feedback before running a grand test.
Testing this will require quite a bit of time and effort which would be wasted if the review results in "uhh, that parallel transform is a bit dodgy after all, let's just stick with sequential processing" 😄

File: src/geocoder.js (hunk 5 above)
Comment (Caerostris) [inline src/geocoder.js]: Ah, I should be able to get rid of the timeout entirely

Comment (ro-ka) [inline src/geocoder.js]: That would be great.

**Comment by ro-ka** [comment]
Added a not to the first entry.  😃

**Comment by Caerostris** [comment]
Travis still isn't accepting the updated `.eslintrc`.
I have tested the rate limitation code under load now. It seems to perform well. ~51QPS including the speedup of the cache with a dataset of 20,000 addresses.
Unfortunately, it is practically impossible to use the full 50 QPS without running into query limits.
I am suspecting that sudden changes in network latency can result in requests reaching the Google servers in uneven intervals.This issue would therefore not be an algorithmic bug, and affect the current master branch as much as this one. The master branch is just so slow that it never runs into this issue.
I have implemented a "retry once" policy for query limit errors in this PR which significantly reduces the number of query limit errors. My test set of 20,000 addresses resulted in one query limit making it into the generated output file.
My proposition is to change the default QPS to a safe number (~35), and add a notice about latency issues with higher QPS to the docs. If desired I could also change the retry policy to a have a configurable number of retries.

**Comment by ro-ka** [comment]
Finding a fast enough value that works is fine with me — if someone wants to go all in, they can define so. Would be ideal if the component just works when using, not running in these issues.

Don’t know whats up with the eslint. Couldn’t reset the cache in Travis.

**Comment by Caerostris** [comment]
That's a 🏆

**Comment by aemkei** [comment]
@ro-ka Should we merge this?

**Comment by Caerostris** [comment]
Yup.

File: src/defaults.js [hunk 6]
```diff
@@ -1,4 +1,7 @@
 export default {
   maxQueriesPerSecond: 50,
-  minQueriesPerSecond: 1
+  minQueriesPerSecond: 1,
+  defaultQueriesPerSecond: 35,
```
Comment (aemkei) [inline src/defaults.js:4]: You should add more infos about this to the README.

Comment (Caerostris) [inline src/defaults.js:4]: Ah, right, you mean why the default is 35 and not 50?

Comment (aemkei) [inline src/defaults.js:4]: Yes, adding a hint, that you can't set the QPS to 50 would be good here:

https://github.com/ubilabs/node-geobatch/pull/12/files/6d35c524cd74a4df282f22e734f5b34309e35d28#diff-04c6e90faac2675aa89e2176d2eec7d8L166

**Comment by aemkei** [comment]
@ro-ka: I've merged that but you need to push it to NMP, cause I "do not have permission to publish "geobatch".

**Comment by ro-ka** [comment]
Published in npm. And @aemkei, you have access, too.  🎉

**Pull request merged**